8-application-demos/6-kalshi-bet-predictor/main.py

```diff
@@ -0,0 +1,170 @@
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer
+from typing import Tuple
+import requests
+import re
+from dotenv import load_dotenv
+import os
+from exa_py import Exa
+
+def getKalshiQuestion(market_ticker)->Tuple[str,str]:
+    url = f"https://api.elections.kalshi.com/trade-api/v2/markets/{market_ticker}"
+    try:
+        res = requests.get(url)
+        res.raise_for_status()
+        obj = res.json()
+        return obj['market']['rules_primary']
+    except requests.exceptions.RequestException as e:
+        raise RuntimeError(f"Error fetching Kalshi market data: {e}")
+
+def getKalshiOdds(market_ticker)->Tuple[str, str]:
+    url = f"https://api.elections.kalshi.com/trade-api/v2/markets/{market_ticker}"
+    try:
+        res = requests.get(url)
+        res.raise_for_status()
+        obj = res.json()
+        return obj['market']['yes_ask'], obj['market']['no_ask']
+    except requests.exceptions.RequestException as e:
+        raise RuntimeError(f"Error fetching Kalshi market data: {e}")
+
+
+class BetPredictor:
+    def __init__(self, model_name: str = "Qwen/Qwen3-4B-Instruct-2507"):
```

Comment:
What made you go with this model out of curioisity?

Reply:
The 3B-6B model size was enough to produce reasonable responses and its really only Qwen and Llama variants in that range. The instruct models are fine tuned for instruction following, and the Qwen3 instruct 2507 model preforms better on instruction following then Llama 3.2 instruct model.